### DISCUSSION TITLE
Testing with a Cloudflare waiting room enabled

### DISCUSSION THREAD

**Discussion by DominicWatson** [artilleryio/artillery#3325] [Artillery]
Is it possible to script a scenario where a request will be retried when encountering either rate limiting or something like a Cloudflare waiting room response?

Currently in my scenario, the VU will just fail and exit the scenario early when running into these - I'd like it to keep retrying in order to emulate a real user attempting to get through the task they want to achieve.

My scenario is just a single flow that is an array of simple GETs and POSTs. On any one of those, I'd like the request to be retried after a short pause IF they encounter a certain response code.

**Comment by hassy** [comment]
Yes should be possible with a bit of custom JS if not something else - what would the check for whether the waiting room is active or not look like?

**Reply by DominicWatson** [reply]
It would be a check on the http response code. i.e. 

* A `200`/`30x` response code would proceed as normal
* A `429` response code would trigger a retry of the request after a 10 second pause to emulate the self refreshing waiting room page
* Any failures would exit the scenario as they currently do (but perhaps they only currently fail due to the capture requirements..)

I'm just looking at custom js possibilities now but starting from scratch.

**Reply by hassy** [reply]
Looks straightforward enough - could a 429 be returned for any request in the scenario?

**Reply by DominicWatson** [reply]
As a brutal try, I'm trying this:

```js
var request = require('sync-request');

module.exports = {
    waitingRoomCheck: waitingRoomCheck
}

function waitingRoomCheck(context, ee, next) {
	request( "GET", context.vars.target, {
		  retry       : true
		, retryDelay  : 10000
		, maxRetries  : 30
	} );

	return next();
}
```

```yml
...
beforeScenario: "waitingRoomCheck"
```

It _could_ be any request, but once you are through the waiting room, you should be able to proceed as normal - so doing this at the beginning of the scenario should work hopefully :)

**Reply by hassy** [reply]
Yep, that should work! Just having a look at what it would take to add the ability to retry on any request - not quite as straightforward as I would've hoped but should be doable as a first-class feature for the next release 🤞 It's a pretty common scenario!

**Reply by DominicWatson** [reply]
Fab, that would be great :). This totally works btw and my load scenario working a treat.

**Reply by DominicWatson** [reply]
This is actually proving a little trickier and the solution I provided above does not work because the session cookies in the request are not persisted to the scenarios VUsers. I'm currently using something like this with some success:

```yml
flow:
  # first loop a request to homepage while responsecode is our waiting room code
  - loop:
      - get:
          url: '/?c={{ $loopCount }}'
          afterResponse: captureResponseCode
      - think: 5
    whileTrue: rateLimitCheck
  
 # then carry on with our flow as expected
  get:
     url: ...
```
*Javascript*
```js
module.exports = {
    rateLimitCheck: rateLimitCheck,
    captureResponseCode: captureResponseCode
}

function rateLimitCheck( context, next ) {
	return next( Number( context.vars.rateLimitCheck ) === 202 ); // I configured the waiting room to return 202
}

function captureResponseCode( req, res, context, ee, next ) {
	context.vars.rateLimitCheck = res.statusCode;
	return next();
}
```

I'm wondering whether or not there is a cleaner way to do the capture. I couldn't find any documented way to do a `capture` in the yaml and get the response status code.
